### PR TITLE
chore: added a new go-i18n-linter

### LINKED ISSUE #1076: Create a custom linter which can detect if a i18n message ID does not exist.
I think it is entirely possible for us to create a custom linter which can be included as part of `golanci-lint` which can detect missing i18n message IDs and fail the build.

When `MustLocalize` or `MustLocalizerError` are used, the linter would look up the i18n locales directory and try to find the ID. If it does not exist, the linter would throw and error.

We can create a private linter by following these docs: https://golangci-lint.run/contributing/new-linters/#how-to-add-a-private-linter-to-golangci-lint

### PR DESCRIPTION
**Pull request by alexal**
Closes #1076 

### Verification Steps
1. Run `make lint`
2. Verify the output. As of right now, there should be three strings that doesn't exist in *.toml files

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation change
- [ ] Other (please specify)

### Checklist

- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

As I already mentioned here https://github.com/redhat-developer/app-services-cli/issues/1076#issuecomment-925750071 the plugin is definitely working once I use it with golangci-lint that I built locally. 

@craicoverflow, @wtrocki I need your help with finding a solution for that.

### CODE REVIEW

**Comment by wtrocki** [comment]
I will look into this in +2 days. Great work!

**Comment by alexal** [comment]
> I will look into this in +2 days. Great work!

Thanks, keep me posted.

**Comment by craicoverflow** [comment]
You should also include the source code of the linter so that we can update/maintain it in the future.

We really need to add a `./hack` folder for this.

**Comment by alexal** [comment]
> You should also include the source code of the linter so that we can update/maintain it in the future.
> 
> We really need to add a `./hack` folder for this.

The source code is available here https://github.com/alexal/go-i18n-linter. Do you want to move it inside app-services-cli project? I also do not mind if we move it inside redhat -developer project. Let me know what you think.

**Comment by alexal** [comment]
> Apologies, I see the link to the external source code now.
No problem, but still, if you want to move that project somewhere, let me know.

**Comment by craicoverflow** [comment]
I think that for the time being it will be better if this is in this repo. Since this works exclusively with our own go-i18n packages it is unlikely to be used outside of this project. If we find a bug in the linter we have no control over it to get a fix in. So all in all, it would be preferable within this repo (under `./hack/linters`).

**Comment by alexal** [comment]
Make sense.

**Comment by alexal** [comment]
> it would be preferable within this repo (under ./hack/linters).

Done

**Comment by wtrocki** [comment]
@alexal Have been doing open source for a while and this tops the best contribution I have seen so far. 
Codebase for linter is very clean and it just works! 
I would have some suggestion before we move forward with merge:

Transfer library to redhat developers while keeping you still as main maintainer of the library.
I guarantee you having access to the codebase for lifetime and us simply wanting to have more control over it for maintenance.

WDYT? Otherwise, this is great to try out on main/while doing development.

**Comment by wtrocki** [comment]
This makes me think as well that this linter can also print out all keys for us in debug mode:

- this will make it easy to create new toml files
- it will simplify checking if we have unused toml keys (still manual but we will have all keys from the code listed)

**Comment by alexal** [comment]
@wtrocki ,

>> Transfer library to redhat developers while keeping you still as main maintainer of the library.
>> I guarantee you having access to the codebase for lifetime and us simply wanting to have more control over it for maintenance.

Sure, I do not mind. Do you want to transfer the library to redhat-developer organization? If yes, I would need your help as I do not have access to it.

>> This makes me think as well that this linter can also print out all keys for us in debug mode

I can do that.

**Comment by wtrocki** [comment]
Let's add me ( `wtrocki` ) to your repo. No further actions are needed.

**Comment by alexal** [comment]
@wtrocki I've sent you an invitation. In this case, I will remove hacks directory from here.

**Comment by wtrocki** [comment]
I needed to be admin so I have actually created new repo with exact code.

https://github.com/redhat-developer/app-services-go-linter

I think we can swap repository here and get it merged. This is mindblowing work!

File: hack/linters/go-i18n-linter/go.sum [hunk 1]
```diff
@@ -0,0 +1,41 @@
+github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
```
Comment (wtrocki) [inline hack/linters/go-i18n-linter/go.sum:1]: We can remove hack folder.

**Comment by wtrocki** [comment]
Sounds like we good! I had long day so will merge that Monday morning to make sure that I did not missed anything

**Comment by alexal** [comment]
> Sounds like we good! I had long day so will merge that Monday morning to make sure that I did not missed anything

Sure. One question though, have you had a success running this linter without compiling your own version of golangci-lint?

**Comment by wtrocki** [comment]
> Sure. One question though, have you had a success running this linter without compiling your own version of golangci-lint?

Why would that be needed.

**Comment by craicoverflow** [comment]
> > Sure. One question though, have you had a success running this linter without compiling your own version of golangci-lint?
> 
> Why would that be needed.

See previous thread - https://github.com/redhat-developer/app-services-cli/issues/1076#issuecomment-925750071

**Comment by wtrocki** [comment]
Yep. I think we need to settle with single relase/hash of golangcilint binary. What we use in github actions vs what plugin is build upon

**Comment by craicoverflow** [comment]
I think @alexal was having problems with the locally installed binary though - not GitHub Actions.

But yes, we should still do that.

**Comment by wtrocki** [comment]
GIthub actions are just example.
Fastest way for us will be to include golangcilint as dependency and run it this way.  
It will slow down cicd but will make things work

**Comment by wtrocki** [comment]
I would try some hacks from stackoverflow like 
https://forum.heroiclabs.com/t/plugin-was-built-with-a-different-version-of-package-internal-unsafeheader/1210

**Comment by alexal** [comment]
> See previous thread - #1076 (comment)

Thanks, @craicoverflow, for adding that comment, I was busy over the weekends.

> I would try some hacks from stackoverflow like

Let me check.

**Comment by craicoverflow** [comment]
@alexal you should install the specific version of golangci-lint which matches the installed Go version:

```shell
curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.42.1
```

When I did this, those errors disappeared, although I got different unrelated errors.

**Comment by alexal** [comment]
@craicoverflow it didn't work for me

image

Because they compile binaries with `CGO_ENABLED=0`
https://github.com/golangci/golangci-lint/blob/cc0914ffeb109a0fca46158081ac8ee0950c71e7/.goreleaser.yml#L32

**Comment by craicoverflow** [comment]
What happens if you also do the same?

**Comment by alexal** [comment]
> What happens if you also do the same?

I won't be able to compile a plugin

image

**Comment by wtrocki** [comment]
I will look into that. Let's leave that for now. Plugin should be build using CI/CD and should be also tested as part of the release.

**Comment by wtrocki** [comment]
@alexal Things were quite busy on our side, but I did not forget about this PR. We going to integrate your linter eventually. 
It is just that we need to get thru deadlines right now

**Comment by wtrocki** [comment]
@rkpattnaik780 if you have time - that will be very useful for us to have in the CLI

**Comment by wtrocki** [comment]
The eagle has landed! Apologies for really long time for getting that merged. This is the most awesome contribution I seen for years! We definitely going to keep it maintained as it is very useful and it is tracking missing keys already.

**Comment by alexal** [comment]
Thank you, @wtrocki. I appreciate your feedback!

**Comment by wtrocki** [comment]
Btw. I have created separate PR that had that change merged. I could not reuse your branch

**Comment by alexal** [comment]
That is fine ;-)

**Pull request closed without merging**